src/full/Agda/TypeChecking/Monad/MetaVars.hs

```diff
@@ -414,6 +414,25 @@ constraintMetas = \case
     listenerMetas EtaExpand{}           = return Set.empty
     listenerMetas (CheckConstraint _ c) = constraintMetas (clValue $ theConstraint c)
 
+-- | Compute all unsolved metas appearing in a set of metas and their solutions. For blocked terms,
+--   return both the meta and metas appearing in the term. Used by generalization to compute
+--   constrained metas. #5845 was caused by not looking at metas appearing in blocked terms.
+unsolvedMetasIn :: Set MetaId -> TCM (Set MetaId)
+unsolvedMetasIn xs = loop Set.empty (Set.toList xs) Set.empty
+  where
+    loop :: Set MetaId -> [MetaId] -> Set MetaId -> TCM (Set MetaId)
```

Comment:
Could you document the purpose of each of these three collections of metas?
That would help to grasp the logic.